### PR TITLE
Add GPU implementation for tf.segment_sum.

### PR DESCRIPTION
**Pull request by tjingrant**
As per #11228, a GPU version of segment sum has been created. 
## Performance 
On a Tesla K40c GPU, the following performance tests have been performed. 
- The test case sizes are represented using 3-tuples of integers. The first two integers correspond to the outer and inner dimension of the input data and the last integer corresponds to the outer dimension of the output data, which is also the total number of segments. 
- The first three rows of data are obtained by running reductions on float32 input type whereas the last three rows of data are obtained by running reductions on float64 input type. 
- We use the GPU version of unsorted segment reduction as the baseline. 
- During experiments, they share the same input data and outputs are compared for consistency.

|  | (1024, 1024, 128)            |   (2048, 2048, 256)          |    (4096, 4096, 512)         | 
|-------------------------------------------------------------|-------------|-------------|-------------| 
| t_unsorted                                                  | 84.942us    | 332.12us    | 1.3170ms    | 
| t_sorted                                                    | 71.047us    | 264.16us    | 1.0391ms    | 
| speedup                                                     | 1.20  | 1.26 | 1.27  | 
|                                                             |             |             |             | 
| t_unsorted                                                  | 160.69us    | 594.93us    | 2.2895ms    | 
| t_sorted                                                    | 106.94us    | 395.19us    | 1.5662ms    | 
| speedup                                                     | 1.50 | 1.51 | 1.46 | 

## Known Limitations:
- We do not check segment_id to make sure that they are within [0, num_outputs-1].
- We do not verify that segment_id is increasing.

### CODE REVIEW

**Comment by tjingrant** [comment]
I signed it!

File: tensorflow/core/kernels/segment_reduction_ops.cc
Comment (zheng-xq) [inline tensorflow/core/kernels/segment_reduction_ops.cc]: This introduces a device/host sync. TensorFlow tries very hard to avoid that. 

An example on how to address such issue: 

https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/where_op.cc#L253

Another common practice is to check the indices on GPU, and do nothing with invalid indicies.

File: tensorflow/core/kernels/segment_reduction_ops.cc
Comment (zheng-xq) [inline tensorflow/core/kernels/segment_reduction_ops.cc]: The convention is to overload "SegmentReductionOp" for GPUDevice

File: tensorflow/core/kernels/segment_reduction_ops.cc
Comment (zheng-xq) [inline tensorflow/core/kernels/segment_reduction_ops.cc]: Since the functor is trivial, we don't have to cache it.

File: tensorflow/core/kernels/segment_reduction_ops_gpu.cu.cc
Comment (zheng-xq) [inline tensorflow/core/kernels/segment_reduction_ops_gpu.cu.cc]: This produces the wrong answer when x == 0. The typical way is to write (x + y - 1) / y. 

In TF, the convention is to use Eigen::divup

File: tensorflow/core/kernels/segment_reduction_ops_gpu.cu.cc
Comment (zheng-xq) [inline tensorflow/core/kernels/segment_reduction_ops_gpu.cu.cc]: Same here. Use div up

**Review by zheng-xq — COMMENTED** [review]
First wave of comments. I'll take a closer look at the details later.

Overall, please add functional tests and benchmarks to make sure it is actually faster across a range of different shapes with different aspect ratios.

Also we need to add more data shapes and randomly generated data to make sure there is no hidden race conditions in the logic.

File: tensorflow/core/kernels/segment_reduction_ops.cc
Comment (tjingrant) [inline tensorflow/core/kernels/segment_reduction_ops.cc]: Hi, the current implementation of SegmentReductionOps is actually rather specialized for non-GPU devices. I'm not sure if the usual different-functor-for-different-device approach is the best way to go here. I actually started this implementation by porting existing code to be reused on GPU but later gave up because it's very inefficient for GPU device; in addition, the original implementation takes data input as non-primitive data types (i.e., TensorMap) which means if I were to use GPU specialized functors, I may have to introduce some extra data types into the .cu file and risk contaminating the .cu file with unnecessary header files; again, with these considerations, I'm not sure overloading SegmentReductionOp is the best approach. Nevertheless, I'm more than happy to consult with your thoughts on this.

File: tensorflow/core/kernels/segment_reduction_ops.cc
Comment (tjingrant) [inline tensorflow/core/kernels/segment_reduction_ops.cc]: Thanks! Will update that.

File: tensorflow/core/kernels/segment_reduction_ops_gpu.cu.cc
Comment (tjingrant) [inline tensorflow/core/kernels/segment_reduction_ops_gpu.cu.cc]: Thanks for picking up this mistake.

File: tensorflow/core/kernels/segment_reduction_ops_gpu.cu.cc
Comment (tjingrant) [inline tensorflow/core/kernels/segment_reduction_ops_gpu.cu.cc]: Will do.

File: tensorflow/core/kernels/segment_reduction_ops_gpu.cu.cc
Comment (zheng-xq) [inline tensorflow/core/kernels/segment_reduction_ops_gpu.cu.cc]: I took a closer look at this code. I think we should definitely add more tests to make sure there is no hidden race condition. Also adding a benchmark showing there is no performance regression against unsorted_segment_sum is good. 

Please make sure try a few different tensor shape aspect ratio in the benchmarks.

File: tensorflow/core/kernels/segment_reduction_ops_gpu.cu.cc
Comment (zheng-xq) [inline tensorflow/core/kernels/segment_reduction_ops_gpu.cu.cc]: As for "MIN", tf convention is "min"

File: tensorflow/core/kernels/segment_reduction_ops_gpu.cu.cc
Comment (zheng-xq) [inline tensorflow/core/kernels/segment_reduction_ops_gpu.cu.cc]: Integer division is quite expensive, maybe pass them in.

**Comment by tjingrant** [comment]
I have updated the pertinent code. Performance tests with more aspect ratios are conducted and recorded in this [gist](https://gist.github.com/tjingrant/3113562d6e9fc8738487ca91847ffd16). The results of both segment_sum and unsorted_segment_sum are always compared for consistency and therefore this also serves as a correctness test for potential race conditions.

In general, the execution time of sorted segment sum cuda kernel is uniformly shorter than that of unsorted segment sum cuda kernel when the problem size is medium or large. It is clear that tiling comes with an inherent cost and therefore when the problem size is small, the performance penalty imposed by tiling will be more noticeable but hopefully nevertheless small. Of all the problem sizes tested (small and large problem sizes alike), the average speed up factor is 1.17 for fp32 and 1.38 for fp64. In general, for fp32 typed segment summation, the break even point is around 512x512 and for any problem sizes beyond this point, using sorted segment sum will be profitable. For fp64, this point is very low and we actually did not see executions of any problem sizes during which sorted segment sum is slower.

It is also worth noting that sorted segment sum requires stream synchronization in order to retrieve output size (a scalar) from device memory. Therefore the overall break even point is somewhat higher but since this cost is constant, it will not impact the execution time of kernels doing more significant work. Please let me know if there is any way to mitigate this problem. Specifically, 

- is it possible to access the input array on the host before it is moved to the device? 
- Or slightly worse, is it possible to have duplicate copies of array at both host and device memory?

**Comment by rmlarsen** [comment]
@zheng-xq can you please take another look?

**Comment by rmlarsen** [comment]
@tensorflow-jenkins test this please

**Comment by tjingrant** [comment]
Hi, judging from the console output, I'm not sure if the test failure is caused by my changes.

`//bazel_pip/tensorflow/contrib/timeseries/python/timeseries:state_management_test FAILED in 13.3s`

This does not look like anything I've touched.

**Comment by rmlarsen** [comment]
@tjingrant indeed, that appears unrelated to your change.

**Comment by rmlarsen** [comment]
@zheng-xq XQ, this appears to pass (except for known flake). Can you approve it?

**Comment by zheng-xq** [comment]
LGTM

@tjingrant, I think it is a good idea to check in your benchmark so the changes are self-sufficient. But I will leave it up to you whether you want to make it into a follow-up CL.

**Comment by tjingrant** [comment]
Updated with benchmarks included.

**Comment by tjingrant** [comment]
ping @zheng-xq , @rmlarsen

**Comment by nolanliou** [comment]
I do a quick test about your code with the [script](https://gist.github.com/nolanliou/a1dc1df3d7d6267a0177b6749ccdf696). the result show there is not as good as unsorted_segment_sum with larger scale of segment_ids.
```
unsorted_segment_sum: 5.616264s
sorted_segment_sum: 7.558630s
unique ids: 6287
```

**Comment by tjingrant** [comment]
Hi @nolanliou , thanks for checking out my code! As for your observation, as you probably know, there are a lot of factors that could affect the performance and after all, your ratio of reductions performed to your total data footprint is quite low and therefore in this case I do not expect my code to perform well. I encourage you to explore more problem sizes and reduction to data footprint ratio. Apart from that, the **main reason** you are seeing such a dramatic performance discrepancy is that you are not using the unsorted segment reductions correctly and probably it's not producing the correct results. Try placing it on a CPU device and you will see the error messages.

File: tensorflow/python/kernel_tests/segment_reduction_ops_test.py
Comment (zheng-xq) [inline tensorflow/python/kernel_tests/segment_reduction_ops_test.py]: Minor nitpicking: maybe np_type_to_str?

File: tensorflow/python/kernel_tests/segment_reduction_ops_test.py
Comment (zheng-xq) [inline tensorflow/python/kernel_tests/segment_reduction_ops_test.py]: Check "is_gpu_available"?

File: tensorflow/python/kernel_tests/segment_reduction_ops_test.py
Comment (zheng-xq) [inline tensorflow/python/kernel_tests/segment_reduction_ops_test.py]: Most benchmarks would split them into two different sub-benchmarks. They can share the underlying implementation

File: tensorflow/python/kernel_tests/segment_reduction_ops_test.py
Comment (zheng-xq) [inline tensorflow/python/kernel_tests/segment_reduction_ops_test.py]: Please check large size with less alignments as well.

File: tensorflow/python/kernel_tests/segment_reduction_ops_test.py
Comment (zheng-xq) [inline tensorflow/python/kernel_tests/segment_reduction_ops_test.py]: We really care about the speedup before and after your CL. Please include the comparison before and after in your commit description. Thanks!

**Review by zheng-xq — CHANGES_REQUESTED** [review]

File: tensorflow/python/kernel_tests/segment_reduction_ops_test.py
Comment (tjingrant) [inline tensorflow/python/kernel_tests/segment_reduction_ops_test.py]: Hi @zheng-xq, there's no (GPU) implementation before my commit. This commit enables segment_sum on GPU. The speedup to the base-line here is another thing. Can you double-check this (the speed up against baseline, not impl before my commit) is what you want?

Comment (zheng-xq) [inline tensorflow/python/kernel_tests/segment_reduction_ops_test.py]: Fair point. In that case, the comparison report in this issue is good enough. If you can include in the commit description, that is helpful. The benchmarks should be separate though.

**Comment by tjingrant** [comment]
Hi @zheng-xq , I think I resolved all your requested changes.

File: tensorflow/python/kernel_tests/segment_reduction_ops_test.py
Comment (zheng-xq) [inline tensorflow/python/kernel_tests/segment_reduction_ops_test.py]: Most benchmark comparison uses percentage in the difference. But it's up to you. Also mention "positive means the new implementation is faster".

Given some kernels do get slower. Add a TODO to investigate why that is the case. 

Thanks for your contribution!

**Comment by zheng-xq** [comment]
LGTM, after minor changes.

**Comment by tjingrant** [comment]
I have modified the commit message to use percentage differences and a todo note is added to the file segment_reduction_ops.cc.

**Comment by zheng-xq** [comment]
LGTM

**Review by zheng-xq — APPROVED** [review]

**Comment by tjingrant** [comment]
Hi @jhseu , any plan to move forward?

**Comment by martinwicke** [comment]
Jenkins, test this please.

**Comment by tjingrant** [comment]
Hi, the error message seems unrelated to my changes:
```
00:36:53 The following tests FAILED:
00:36:53 	252 - C:/tf_jenkins/home/workspace/tensorflow-pr-win-cmake-py/tensorflow/python/training/server_lib_test.py (Timeout)
00:36:53 Errors while running CTest
```

**Pull request merged**
